File: Semester3/MAP/Seminare/Sem10/Sem10/Decorator/AbstractTaskRunner.cs
namespace Sem10;

public abstract class AbstractTaskRunner : TaskRunner
{
    protected TaskRunner runner;

    public AbstractTaskRunner(TaskRunner runner)
    {
        this.runner = runner;
    }

    public virtual void executeOneTask()
    {
        runner.executeOneTask();
    }

    public void executeAll()
    {
        while(runner.hasTasks())
            executeOneTask();
    }

    public void addTask(Task task)
    {
        runner.addTask(task);
    }

    public bool hasTasks()
    {
        return runner.hasTasks();
    }
}

File: Semester3/MAP/laboratoare/lab2/TaskManager/src/Decorator/AbstractTaskRunner.java
package Decorator;

import Domain.Task;

public abstract class AbstractTaskRunner implements TaskRunner {
    private TaskRunner taskRunner;

    public AbstractTaskRunner(TaskRunner taskRunner) {
        this.taskRunner = taskRunner;
    }

    @Override
    public void executeOneTask() {
        taskRunner.executeOneTask();
    }

    @Override
    public void executeAll() {
        while(hasTask())
            executeOneTask();
    }

    @Override
    public void addTask(Task task) {
        taskRunner.addTask(task);
    }

    @Override
    public boolean hasTask() {
        return taskRunner.hasTask();
    }
}
